### PR TITLE
implement logging timestamps and record container ID in a file

### PR DESCRIPTION
**Pull request by MichaelTong**
Added options:
```
--record-container-id
                      If enabled, a file with suffix ".cid" will be created
                      storing the container ID under CIDFILE_DIR
--cidfile-dir CIDFILE_DIR
                      Directory for storing cidfiles. Default at /tmp/
--cidfile-prefix CIDFILE_PREFIX
                      Give a prefix to cidfile. Final file name will be
                      followed by a timestamp. Default empty.
--logtstp             Print timestamps with logging
```

### CODE REVIEW

**Comment by mr-c** [comment]
Jenkins, okay to test

**Comment by MichaelTong** [comment]
Not sure why failed Jenkins... Console Output didn't give much information..

File: cwltool/main.py [hunk 1]
```diff
@@ -84,6 +84,20 @@ def arg_parser():  # type: () -> argparse.ArgumentParser
                          default=True, help="Do not delete Docker container used by jobs after they exit",
                          dest="rm_container")
 
+    parser.add_argument("--record-container-id", action="store_true", default=False,
+                        help="If enabled, a file with suffix \".cid\" will be created storing the container ID under CIDFILE_DIR",
+                        dest="record_container_id")
+
+    parser.add_argument("--cidfile-dir", type=Text,
+                        help="Directory for storing cidfiles. Default at /tmp/",
+                        default="/tmp/",
```
Comment (mr-c) [inline cwltool/main.py]: I think the current directory is a better default and is more portable.

File: cwltool/main.py [hunk 2]
```diff
@@ -84,6 +84,20 @@ def arg_parser():  # type: () -> argparse.ArgumentParser
                          default=True, help="Do not delete Docker container used by jobs after they exit",
                          dest="rm_container")
 
+    parser.add_argument("--record-container-id", action="store_true", default=False,
+                        help="If enabled, a file with suffix \".cid\" will be created storing the container ID under CIDFILE_DIR",
```
Comment (mr-c) [inline cwltool/main.py]: Can you linewrap at 80 for easier reviews?

Comment (mr-c) [reply cwltool/main.py]: "storing the container ID under the directory specified by --cidfile-dir"

Comment (mr-c) [reply cwltool/main.py]: These new options would be best to put in their own https://docs.python.org/2.7/library/argparse.html#argument-groups. You can use that help to explain what a "cidfile" is.

File: cwltool/main.py [hunk 3]
```diff
@@ -161,6 +175,7 @@ def arg_parser():  # type: () -> argparse.ArgumentParser
     exgroup.add_argument("--verbose", action="store_true", help="Default logging")
     exgroup.add_argument("--quiet", action="store_true", help="Only print warnings and errors.")
     exgroup.add_argument("--debug", action="store_true", help="Print even more logging")
+    parser.add_argument("--logtstp", action="store_true", help="Print timestamps with logging")
```
Comment (mr-c) [inline cwltool/main.py]: I think `--timestamps` would be a better name and "Add timestamps to the errors, warnings, and notifications."

File: cwltool/job.py [hunk 4]
```diff
@@ -463,6 +466,25 @@ def run(self, pull_image=True, rm_container=True,
...
         # runtime.append("--env=HOME=/tmp")
         runtime.append(u"--env=HOME=%s" % self.builder.outdir)
+        
+        # add parameters to docker to write a container ID file
+        if record_container_id:
+            if cidfile_dir:
+                if not os.path.isdir(cidfile_dir):
+                    _logger.error("--cidfile-dir %s error:\n%s", cidfile_dir,
+                                  cidfile_dir + " is not a directory or directory doesn't exist, please check it first")
+                    exit(-1)
```
Comment (mr-c) [inline cwltool/job.py]: Should be `exit(2)`: https://docs.python.org/2/library/sys.html#sys.exit

File: cwltool/job.py [hunk 5]
```diff
@@ -463,6 +466,25 @@ def run(self, pull_image=True, rm_container=True,
...
+                if not os.path.exists(cidfile_dir):
+                    _logger.error("--cidfile-dir %s error:\n%s", cidfile_dir,
+                                  "directory doesn't exist, please create it first")
+                    exit(-1)
```
Comment (mr-c) [inline cwltool/job.py]: Should be `exit(2)`: https://docs.python.org/2/library/sys.html#sys.exit

File: cwltool/job.py [hunk 6]
```diff
@@ -463,6 +466,25 @@ def run(self, pull_image=True, rm_container=True,
...
+            else:
+                cidfile_dir = "/tmp/"
+            cidfile_name = datetime.datetime.now().strftime("%Y%m%d%H%M%S-%f")+".cid"
+            if cidfile_prefix != "":
+                cidfile_name = str(cidfile_prefix + "-" + cidfile_name)
+            cidfile_path = cidfile_dir + cidfile_name
+            runtime.append(u"--cidfile=%s" % cidfile_path)
```
Comment (mr-c) [inline cwltool/job.py]: Is there an equivalent for Singularity?

**Comment by mr-c** [comment]
Jenkins, test this please

**Comment by mr-c** [comment]
Jenkins, test this please

File: cwltool/job.py (hunk 6 above)
Comment (MichaelTong) [reply cwltool/job.py]: Sorry but I don't think I understand this. Can you explain it a little bit.
I have finished other suggested changes

Comment (mr-c) [reply cwltool/job.py]: http://singularity.lbl.gov/

If this option only works for the Docker runtime then we should say so. If we can add the equivalent option for Singularity then that would be nice.

Comment (MichaelTong) [reply cwltool/job.py]: As far as I know, it comes with docker. I didn't know singularity before, but by reading its documents, I believe it's possible to add the equivalent option.

**Pull request merged**